src/tools/tidy/src/fluent_checks.rs

```diff
@@ -34,16 +39,58 @@ run `./x.py test tidy --bless` to sort the file correctly",
     }
 }
 
-fn sort_messages(fluent: &str) -> String {
+fn check_unused_messages(
+    filename: &str,
+    fluent: &str,
+    bad: &mut bool,
+    report: bool,
+) -> Vec<String> {
+    let mut matches = MESSAGE.captures_iter(fluent).peekable();
+    let mut sources: Vec<String> = vec![];
+    if matches.peek().is_some() {
+        let src_dir = Path::new(filename).parent().unwrap().join("src");
+        walk(
+            &src_dir,
+            |path, is_dir| filter_dirs(path) || (!is_dir && filter_not_rust(path)),
+            &mut |_ent, contents| {
+                sources.push(contents.to_string());
+            },
+        );
+    }
+    let mut unused_messages = vec![];
+    while let Some(m) = matches.next() {
+        let name = m.get(1).unwrap();
+        // message maybe referenced as {name} in fluent files
+        let name_var = "{".to_string() + name.as_str() + "}";
+        if !sources.iter().any(|file| file.contains(name.as_str())) && !fluent.contains(&name_var) {
```

Comment:
This feels like it might get quite slow. It seems like we're scanning ~every Rust file in the same compiler crate for *each* fluent string. But, maybe OK for now, I don't think we need to block on it at this time.